### PR TITLE
Consolidate XDR into the mfhdf library

### LINKED ISSUE #15: XDR Undefined References
Hi there,

I'm attempting to build a package for Alpine Linux from v4.2.15.  However, when trying to link in the XDR types, `ld` is failing with many undefined references, for example to `xdr_u_long`.

I initially suspected that this was the incompatibility mentioned in section 4.4 of [this document](https://github.com/HDFGroup/hdf4/blob/master/release_notes/INSTALL), meaning my system's version of xdr (portablexdr) is incompatible with this project.

Therefore I supplied configure with `--enable-hdf4-xdr`, and removed portablexdr from my system.  The bundled version of XDR gets compiled, but I get the same behaviour.

The full linker output can be found [here](https://gist.github.com/eddsalkield/f870d601dfb27052d2c94f8156d1de54).  I'd appreciate any input you can give as to how I might proceed :)

### LINKED ISSUE #25: 4.2.15: cmake build with `HDF4_BUILD_XDR_LIB=OFF` fails on linking
cmake settings:
```console
[tkloczko@devel-g2v x86_64-redhat-linux-gnu]$ cmake -L
CMake Warning:
  No source or binary directory provided.  Both will be assumed to be the
  same as the current working directory, but note that this warning will
  become a fatal error in future CMake releases.


CMake Error: The source directory "/home/tkloczko/rpmbuild/BUILD/hdf-4.2.15/x86_64-redhat-linux-gnu" does not appear to contain CMakeLists.txt.
Specify --help for usage, or press the help button on the CMake GUI.
-- Cache values
BUILD_SHARED_LIBS:BOOL=ON
BUILD_STATIC_LIBS:BOOL=ON
BUILD_TESTING:BOOL=ON
BUILD_USER_DEFINED_LIBS:BOOL=OFF
CMAKE_ARCHIVE_OUTPUT_DIRECTORY:PATH=/home/tkloczko/rpmbuild/BUILD/hdf-4.2.15/x86_64-redhat-linux-gnu/bin
CMAKE_BUILD_TYPE:STRING=RelWithDebInfo
CMAKE_Fortran_MODULE_DIRECTORY:PATH=/home/tkloczko/rpmbuild/BUILD/hdf-4.2.15/x86_64-redhat-linux-gnu/bin
CMAKE_INSTALL_PREFIX:PATH=/usr
CMAKE_Java_ARCHIVE:FILEPATH=/usr/bin/jar
CMAKE_Java_RUNTIME:FILEPATH=/usr/bin/java
CMAKE_LIBRARY_OUTPUT_DIRECTORY:PATH=/home/tkloczko/rpmbuild/BUILD/hdf-4.2.15/x86_64-redhat-linux-gnu/bin
CMAKE_RUNTIME_OUTPUT_DIRECTORY:PATH=/home/tkloczko/rpmbuild/BUILD/hdf-4.2.15/x86_64-redhat-linux-gnu/bin
CTEST_TEST_TIMEOUT:STRING=3600
HDF4_ALLOW_EXTERNAL_SUPPORT:STRING=ON
HDF4_BUILD_EXAMPLES:BOOL=OFF
HDF4_BUILD_FORTRAN:BOOL=OFF
HDF4_BUILD_JAVA:BOOL=ON
HDF4_BUILD_TOOLS:BOOL=ON
HDF4_BUILD_UTILS:BOOL=ON
HDF4_BUILD_XDR_LIB:BOOL=OFF
HDF4_DISABLE_COMPILER_WARNINGS:BOOL=ON
HDF4_ENABLE_COVERAGE:BOOL=OFF
HDF4_ENABLE_DEPRECATED_SYMBOLS:BOOL=OFF
HDF4_ENABLE_JPEG_LIB_SUPPORT:BOOL=ON
HDF4_ENABLE_NETCDF:BOOL=ON
HDF4_ENABLE_SZIP_SUPPORT:BOOL=OFF
HDF4_ENABLE_Z_LIB_SUPPORT:BOOL=ON
HDF4_JAVA_PACK_JRE:BOOL=OFF
HDF4_PACKAGE_EXTLIBS:BOOL=ON
HDF4_PACK_EXAMPLES:BOOL=OFF
HDF_ENABLE_LARGE_FILE:BOOL=ON
JPEG_DIR:PATH=JPEG_DIR-NOTFOUND
XDR_INCLUDE_DIR:PATH=/usr/include/tirpc
XDR_INT_LIBRARY:FILEPATH=/usr/lib64/libtirpc.so
ZLIB_DIR:PATH=ZLIB_DIR-NOTFOUND
```
and linking fails on
```console
[ 28%] Linking C executable ../../bin/hdftest
cd /home/tkloczko/rpmbuild/BUILD/hdf-4.2.15/x86_64-redhat-linux-gnu/mfhdf/test && /usr/bin/cmake -E cmake_link_script CMakeFiles/hdftest.dir/link.txt --verbose=1
/usr/bin/gcc  -std=c99 -O2 -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -fdata-sections -ffunction-sections -flto=auto -flto-partition=none -fstdarg-opt -w -pedantic -Wall -Wextra -fmessage-length=0 -DNDEBUG -Wl,-z,relro -Wl,--as-needed -Wl,--gc-sections -Wl,-z,now -specs=/usr/lib/rpm/redhat/redhat-hardened-ld -flto=auto -flto-partition=none -fuse-linker-plugin -Wl,--build-id=sha1 CMakeFiles/hdftest.dir/hdftest.c.o CMakeFiles/hdftest.dir/tchunk.c.o CMakeFiles/hdftest.dir/tcomp.c.o CMakeFiles/hdftest.dir/tcoordvar.c.o CMakeFiles/hdftest.dir/tdim.c.o CMakeFiles/hdftest.dir/temptySDSs.c.o CMakeFiles/hdftest.dir/tattributes.c.o CMakeFiles/hdftest.dir/tfile.c.o CMakeFiles/hdftest.dir/tmixed_apis.c.o CMakeFiles/hdftest.dir/tnetcdf.c.o CMakeFiles/hdftest.dir/trank0.c.o CMakeFiles/hdftest.dir/tsd.c.o CMakeFiles/hdftest.dir/tsdsprops.c.o CMakeFiles/hdftest.dir/tszip.c.o CMakeFiles/hdftest.dir/tattdatainfo.c.o CMakeFiles/hdftest.dir/tdatainfo.c.o CMakeFiles/hdftest.dir/tdatasizes.c.o CMakeFiles/hdftest.dir/texternal.c.o CMakeFiles/hdftest.dir/tutils.c.o -o ../../bin/hdftest  -Wl,-rpath,"\$ORIGIN/../lib:\$ORIGIN/" ../../bin/libmfhdf.so.4.15.2 ../../bin/libhdf.so.4.15.2
/usr/bin/ld: ../../bin/libmfhdf.so.4.15.2: undefined reference to `xdr_enum'
/usr/bin/ld: ../../bin/libmfhdf.so.4.15.2: undefined reference to `xdr_int'
/usr/bin/ld: ../../bin/libmfhdf.so.4.15.2: undefined reference to `xdr_u_long'
/usr/bin/ld: ../../bin/libmfhdf.so.4.15.2: undefined reference to `xdr_long'
/usr/bin/ld: ../../bin/libmfhdf.so.4.15.2: undefined reference to `xdr_bytes'
/usr/bin/ld: ../../bin/libmfhdf.so.4.15.2: undefined reference to `xdr_opaque'
/usr/bin/ld: ../../bin/libmfhdf.so.4.15.2: undefined reference to `xdr_double'
/usr/bin/ld: ../../bin/libmfhdf.so.4.15.2: undefined reference to `xdr_float'
collect2: error: ld returned 1 exit status
make[2]: *** [mfhdf/test/CMakeFiles/hdftest.dir/build.make:390: bin/hdftest] Error 1
make[2]: Leaving directory '/home/tkloczko/rpmbuild/BUILD/hdf-4.2.15/x86_64-redhat-linux-gnu'
make[1]: *** [CMakeFiles/Makefile2:1567: mfhdf/test/CMakeFiles/hdftest.dir/all] Error 2
make[1]: Leaving directory '/home/tkloczko/rpmbuild/BUILD/hdf-4.2.15/x86_64-redhat-linux-gnu'
make: *** [Makefile:169: all] Error 2
```

### LINKED ISSUE #151: Build fails on x86_64 OpenBSD 7.2 clang
```
In file included from /home/runner/work/hdf4/hdf4/hdf4/mfhdf/libsrc/array.c:18:
[1826](https://github.com/HDFGroup/hdf4/actions/runs/3950638539/jobs/6763440046#step:3:1827)
/home/runner/work/hdf4/hdf4/hdf4/mfhdf/libsrc/local_nc.h:487:62: error: unknown type name 'u_int'
[1827](https://github.com/HDFGroup/hdf4/actions/runs/3950638539/jobs/6763440046#step:3:1828)
HDFLIBAPI bool_t xdr_shorts    PROTO((XDR * xdrs, short *sp, u_int cnt));
[1828](https://github.com/HDFGroup/hdf4/actions/runs/3950638539/jobs/6763440046#step:3:1829)
                                                             ^
[1829](https://github.com/HDFGroup/hdf4/actions/runs/3950638539/jobs/6763440046#step:3:1830)
/home/runner/work/hdf4/hdf4/hdf4/mfhdf/libsrc/array.c:480:11: error: expected ';' after expression
[1830](https://github.com/HDFGroup/hdf4/actions/runs/3950638539/jobs/6763440046#step:3:1831)
    u_long  count = 0, *countp = NULL;
[1831](https://github.com/HDFGroup/hdf4/actions/runs/3950638539/jobs/6763440046#step:3:1832)
          ^
```

### LINKED ISSUE #423: Remove the dependency on XDR
XDR will be removed and replaced with a thin I/O layer that does handles long integer data correctly.

XDR is currently not in the public API and is not otherwise exposed. Users should experience no change in behavior from this change.

### PR DESCRIPTION
**Pull request by derobins**
XDR is now consolidated into the mfhdf library:

* The build system always uses the library XDR implementation
* Unused XDR functionality has been removed
* There is no separate XDR library
* The xdr.h header is no longer deployed
* The POSIX XDR implementation is hard-coded as the only XDR implementation
* The XDR "long" functions have been replaced with functions that take 32-bit data

XDR will eventually be removed from HDF4, but in the meantime, this will make it easier to deal with that layer and will fix the linking issues seen on some platforms.

Fixes #15
Fixes #25 
Fixes #423
Fixes #151

### CODE REVIEW

File: mfhdf/libsrc/file.c [hunk 1]
```diff
@@ -1033,3 +1030,58 @@ ncsetfill(int id, int fillmode)
...
+    }
+    else {
+        op = XDR_DECODE;
+    }
+
+    if (xdrposix_create(xdrs, fd, fmode, op) < 0)
+        return -1;
+    else
+        return fd;
+}
```
Comment (derobins) [inline mfhdf/libsrc/file.c:1087]: These were moved from xdrposix.c so we don't have to pull in the netCDF header in xdr.c

File: mfhdf/libsrc/xdrposix.c
Comment (derobins) [inline mfhdf/libsrc/xdrposix.c:1]: Functionality was moved to xdr.c

File: mfhdf/nctest/nctest.lnk
Comment (derobins) [inline mfhdf/nctest/nctest.lnk:1]: I'd be willing to bet we can delete this .lnk file. I'll do that in a separate PR.

File: mfhdf/test/CMakeTests.cmake [hunk 2]
```diff
@@ -130,44 +130,3 @@ set_tests_properties (MFHDF_TEST-hdfnctest PROPERTIES
...
-            -D "TEST_ARGS:STRING="
-            -D "TEST_FOLDER=${PROJECT_BINARY_DIR}/TEST"
-            -D "TEST_OUTPUT=xdrtest.tst"
-            -D "TEST_EXPECT=0"
-            -D "TEST_REFERENCE=xdrtest.out"
-            -P "${HDF_RESOURCES_DIR}/runTest.cmake"
-    )
-  endif ()
-  set_tests_properties (MFHDF_TEST-xdrtest PROPERTIES DEPENDS hdfnctest LABELS ${PROJECT_NAME})
-endif ()
```
Comment (derobins) [inline mfhdf/test/CMakeTests.cmake:173]: xdrtest tests the stdio flavor of XDR which isn't what the library uses and doesn't even test much. I'm ditching it so I can get rid of xdrstdio and hard-code xdrposix, in preparation for completely removing XDR.

File: mfhdf/xdr/byteordr.c
Comment (derobins) [inline mfhdf/xdr/byteordr.c:1]: This was unused

**Review by bmribler — APPROVED** [review]

**Review by bmribler — APPROVED** [review]

File: config/cmake/ConfigureChecks.cmake [hunk 3]
```diff
@@ -64,7 +64,7 @@ if (WIN32 AND NOT MINGW)
 endif ()
 
 if (WINDOWS)
-  set (HDF4_REQUIRED_LIBRARIES "ws2_32.lib;wsock32.lib")
+  set (CMAKE_REQUIRED_LIBRARIES "ws2_32.lib;wsock32.lib")
```
Comment (byrnHDF) [inline config/cmake/ConfigureChecks.cmake:67]: Why the change?

**Review by byrnHDF — APPROVED** [review]

File: config/cmake/ConfigureChecks.cmake (hunk 3 above)
Comment (derobins) [reply config/cmake/ConfigureChecks.cmake:67]: I was trying to get MinGW to work and this seemed like the more correct thing to do. Is it not?

Comment (derobins) [reply config/cmake/ConfigureChecks.cmake:67]: Isn't that the point, though? We want to require links to ws2_32, etc.

Comment (byrnHDF) [reply config/cmake/ConfigureChecks.cmake:67]: Actually it looks like the opposite is true -  CMAKE_REQUIRED_LIBRARIES is temporary for checking, we use HDF4_REQUIRED_LIBRARIES after the configure_checks.

Comment (byrnHDF) [reply config/cmake/ConfigureChecks.cmake:67]: But then again CMAKE_REQUIRED_LIBRARIES shows up afterwards - ARRGH.

Comment (byrnHDF) [reply config/cmake/ConfigureChecks.cmake:67]: HDF5 uses CMAKE_REQUIRED_LIBRARIES is temporary for checking.

Comment (byrnHDF) [reply config/cmake/ConfigureChecks.cmake:67]: For HDF4 CMakeLists.txt line 485 incorrectly set CMAKE_REQUIRED_LIBRARIES and should have been setting HDF4_REQUIRED_LIBRARIES.
Then everything would be correct ;
CMAKE_REQUIRED_LIBRARIES is temporary for checking, we use HDF4_REQUIRED_LIBRARIES after the configure_checks.

**Pull request merged**